File: src/main/java/tasks/CodeForces/taskD.cpp
#include <bits/stdc++.h>

using namespace std;

class StackMin {
private:
    vector<pair<int, int>> mins;
    vector<int> stack;

public:
    StackMin() {}

    int get_min() {
        return mins.back().second;
    }

    void push(int x) {
        int i = stack.size();
        stack.push_back(x);
        if (i == 0 || mins.back().second > x) {
            mins.push_back(make_pair(i, x));
        }
    }

    void pop() {
        int i = stack.size() - 1;
        stack.pop_back();
        if (mins.back().first == i) {
            mins.pop_back();
        }
    }
};

void solve() {
    int n;
    cin >> n;
    StackMin st;
    for (int j = 0; j < n; j++) {
        char s;
        cin >> s;
        if (s == '1') {
            int x;
            cin >> x;
            st.push(x);
        } else if (s == '2') {
            st.pop();
        } else {
            cout << st.get_min() << endl;
        }
    }
}

int main() {
    solve();
    return 0;
}

File: src/main/java/tasks/CodeForces/taskD.py
class StackMin:
    def __init__(self):
        self.mins = []
        self.stack = []
    def get_min(self):
        return self.mins[-1][1]

    def push(self, x):
        i = len(self.stack)
        self.stack.append(x)
        if i == 0 or self.mins[-1][1] > x:
            self.mins.append((i, x))

    def pop(self):
        i = len(self.stack) - 1
        self.stack.pop()
        if self.mins[-1][0] == i:
            self.mins.pop()


def solve():
    n = int(input())
    st = StackMin()
    for _ in range(n):
        s = input()
        if s[0] == '1':
            x = int(s[2:])
            st.push(x)
        elif s[0] == '2':
            st.pop()
        else:
            print(st.get_min())

solve()
